<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Intro to rescueSim • rescueSim</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Intro to rescueSim">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rescueSim</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/rescueSimPowerVignette.html">Power Analysis with rescueSim</a></li>
    <li><a class="dropdown-item" href="../articles/rescueSimVignette.html">Intro to rescueSim</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ewynn610/rescueSim/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Intro to rescueSim</h1>
            
            <h4 data-toc-skip class="date">June 30, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ewynn610/rescueSim/blob/master/vignettes/rescueSimVignette.Rmd" class="external-link"><code>vignettes/rescueSimVignette.Rmd</code></a></small>
      <div class="d-none name"><code>rescueSimVignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>rescueSim</code> is an R package for simulating single-cell
RNA-sequencing (scRNA-seq) data from repeated measures designs, such as
longitudinal studies. One benefit of these types of studies is they
allow researchers to investigate changes in gene expression over time
for a specific cell type. Realistic simulated data data can be used to
evaluate analysis methods and assess power under a variety of
settings.</p>
<p>The <code>rescueSim</code> package uses a gamma-poisson framework to
simulate data from individual cell-types while mimicking important
attributes of empirical data such as between-sample variability
(i.e. data from cells from the same sample are correlated) and
between-subject variability (i.e. data from samples from the same
subject are correlated).</p>
<p>In this vignette, we outline:</p>
<ul>
<li>An overview of the simulation method</li>
<li>The basic workflow for simulating data</li>
<li>How to generate and update an object for storing simulation
parameters</li>
<li>How to estimate simulation parameters from a data set</li>
<li>How to simulate data using simulation parameters</li>
<li>The effect of adjusting some of the simulation parameters on the
simulated data</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>rescueSim</code> is available on github at: <a href="https://github.com/ewynn610/rescueSim" class="external-link uri">https://github.com/ewynn610/rescueSim</a>.</p>
<p>It can be installed using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"https://github.com/ewynn610/rescueSim"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="simulation-overview">Simulation Overview<a class="anchor" aria-label="anchor" href="#simulation-overview"></a>
</h2>
<p><code>rescueSim</code> simulates a data set for a <em>single</em>
cell type with gene expression counts for cells from several samples,
with multiple samples coming from individual subjects. If a user wants
data from several cell types, this process can be repeated to generate
data for each cell type of interest.</p>
<p>The simulation process can be broken up into three steps (Figure
1):</p>
<ol style="list-style-type: decimal">
<li>Simulate sample-specific mean expression values for each gene, with
values for samples from the same subject having an extra level of
correlation.</li>
<li>Draw cell-level “true” expression values for each gene using the
sample-specific values.</li>
<li>Simulate final count values for each cell with added variability
representing technical effects.</li>
</ol>
<div class="float">
<img src="fig1.png" style="width:75.0%" alt="Figure 1 Basic simulation framework"><div class="figcaption">Figure 1 Basic simulation framework</div>
</div>
<p>As seen in the left hand side of Figure 1, the simulation framework
relies on a series of parameters. These can be estimated from a data set
provided by the user or supplied manually by the user. Alternatively,
predefined parameter values which were estimated from a longitudinal
scRNA-Seq data set of bronchoalveolar lavage samples from healthy adult
subjects are included in the package. Users can use a combination of
predefined or supplied parameters and parameters estimated from
empirical data. This is particularly useful in situations where
multi-sample/multi-time point data is not available in which case some,
but not all, simulation parameters can be estimated from the empirical
data.</p>
</div>
<div class="section level2">
<h2 id="workflow-basics">Workflow Basics<a class="anchor" aria-label="anchor" href="#workflow-basics"></a>
</h2>
<p>The simulation workflow includes two broad steps:</p>
<ol style="list-style-type: decimal">
<li>Generating simulation parameters (using
<code><a href="../reference/RescueSimParams.html">RescueSimParams()</a></code> and
<code>estimateRescueSimParams()</code>)</li>
<li>Using the simulation parameters to simulate the data (using
<code>simulateRescueData()</code>)</li>
</ol>
<p>The end result is a <code>singleCellExperiment</code> object with
simulated data for an <em>individual</em> cell-type. If you want data
from several cell types, you can re-run this process, preferably using
simulation parameters estimated from data from the cell type(s) of
interest.</p>
<p>Throughout the following example we will use gene expression data for
recruited airspace macrophage cells from bronchoalveolar lavage samples
samples collected from healthy adults. The data has been subset to
contain gene expression for 1940 genes from 976 cells. Data was
collected from five subjects at two time points per subject.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load necessary packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ewynn610/rescueSim" class="external-link">rescueSim</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">24</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"RecAM_sce"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Data is held in singleCellExperiment object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SingleCellExperiment"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "SingleCellExperiment"</span></span>
<span></span>
<span><span class="co">## Data dimensions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1940  976</span></span>
<span></span>
<span><span class="co">## Cell level metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 3 columns</span></span>
<span><span class="co">#&gt;                       sampleID   subjectID        time</span></span>
<span><span class="co">#&gt;                    &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; ACTCCCAAGTCATTGC_2     sample1    subject1       time0</span></span>
<span><span class="co">#&gt; TATTGCTTCCTAGCCT_2     sample1    subject1       time0</span></span>
<span><span class="co">#&gt; ACGGGTCTCGTCCTTG_2     sample1    subject1       time0</span></span>
<span><span class="co">#&gt; AACGAAACATGGGCAA_2     sample1    subject1       time0</span></span>
<span><span class="co">#&gt; GAGACCCCACGGTGTC_2     sample1    subject1       time0</span></span>
<span><span class="co">#&gt; GGGAAGTTCTCAACGA_2     sample1    subject1       time0</span></span>
<span></span>
<span><span class="co">## Five subjects</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">$</span><span class="va">subjectID</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "subject1" "subject2" "subject3" "subject4" "subject5"</span></span>
<span></span>
<span><span class="co">## Ten samples (two per subject)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">$</span><span class="va">sampleID</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "sample1"  "sample2"  "sample3"  "sample4"  "sample5"  "sample6" </span></span>
<span><span class="co">#&gt;  [7] "sample7"  "sample8"  "sample9"  "sample10"</span></span>
<span></span>
<span><span class="co">## Two different time points</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">$</span><span class="va">time</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "time0" "time1"</span></span></code></pre></div>
<p>We’ll begin with a smaller data set of just 40 genes to show the
basic structure of the package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subset data to 40 genes</span></span>
<span><span class="va">RecAM_sce_small</span> <span class="op">&lt;-</span> <span class="va">RecAM_sce</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">40</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generating-simulation-parameters">Generating simulation parameters<a class="anchor" aria-label="anchor" href="#generating-simulation-parameters"></a>
</h2>
<p>Objects of the class <code>RescueSimParams</code> are used to hold
simulation parameters and settings. A parameter object can be generated
using the code below.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create params object</span></span>
<span><span class="va">myParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RescueSimParams.html">RescueSimParams</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Param object class</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">myParams</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "RescueSimParams"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "rescueSim"</span></span>
<span></span>
<span><span class="co">## Examine param object contents</span></span>
<span><span class="va">myParams</span></span>
<span><span class="co">#&gt; An object of class "RescueSimParams"</span></span>
<span><span class="co">#&gt; Slot "nTimepoints":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "twoGroupDesign":</span></span>
<span><span class="co">#&gt; logical(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "nSubjsPerGroup":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "maxCellsPerSamp":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "minCellsPerSamp":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibMean":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibSD":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibFacVar":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "customLibSizes":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "exprsMean":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "dispersion":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "sampleFacVarMean":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "sampleFacVarSD":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "subjectFacVarMean":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "subjectFacVarSD":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "propDE":</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "deLog2FC":</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>We can see that the object contains slots for a variety of parameters
which are empty by default. These slots represent the parameters in the
left hand side of figure 1 used to simulate the data as well as other
data properties such as the number of time points or subjects.</p>
<p>The table below gives an details about the simulation parameters.</p>
<table class="table">
<colgroup>
<col width="3%">
<col width="5%">
<col width="33%">
<col width="57%">
</colgroup>
<thead><tr class="header">
<th align="center">Parameter</th>
<th align="center">Symbol in Fig. 1 (if applicable)</th>
<th align="center">Description</th>
<th align="center">Acceptable Values</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><code>nTimepoints</code></td>
<td align="center">-</td>
<td align="center">Number of time points (i.e. samples) per subject</td>
<td align="center">Single numeric value &gt;0 representing the number of
time points for all subjects.</td>
</tr>
<tr class="even">
<td align="center"><code>twoGroupDesign</code></td>
<td align="center">-</td>
<td align="center">Logical value indicating whether to simulate two
groups (ex. treatment and control group)</td>
<td align="center">Logical value with <code>TRUE</code> indicating that
two groups should be simulated.</td>
</tr>
<tr class="odd">
<td align="center"><code>nSubjsPerGroup</code></td>
<td align="center">-</td>
<td align="center">Number of subjects per group (if using two group
design) or number of total subjects (if only single group)</td>
<td align="center">A single numeric value &gt;0 indicating the number of
subjects per group.</td>
</tr>
<tr class="even">
<td align="center"><code>maxCellsPerSamp</code></td>
<td align="center">-</td>
<td align="center">Maximum parameter used when drawing number of cells
per sample from a discrete uniform distribution</td>
<td align="center">A single numeric value &gt;0 indicating the maximum
number of cells per sample, or a vector with length equal to the number
of conditions (group/time point combinations) with each value
representing the maximum for a single condition, or a vector with length
equal to the number of total samples with each value representing the
maximum for a single sample.</td>
</tr>
<tr class="odd">
<td align="center"><code>minCellsPerSamp</code></td>
<td align="center">-</td>
<td align="center">Minimum parameter used when drawing number of cells
per sample from a discrete uniform distribution</td>
<td align="center">A single numeric value &gt;0 indicating the minimum
number of cells per sample, or a vector with length equal to the number
of conditions (group/time point combinations) with each value
representing the minimum for a single condition, or a vector with length
equal to the number of total samples with each value representing the
minimum for a single sample.</td>
</tr>
<tr class="even">
<td align="center"><code>logLibMean</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mi>L</mi></msub><annotation encoding="application/x-tex">\mu_L</annotation></semantics></math></td>
<td align="center">Mean library size (log scale) parameter. Used to draw
library size from a log-normal distribution.</td>
<td align="center">A single value &gt;0 indicating the mean library size
on a log scale.</td>
</tr>
<tr class="odd">
<td align="center"><code>logLibSD</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>σ</mi><mi>L</mi></msub><annotation encoding="application/x-tex">\sigma_L</annotation></semantics></math></td>
<td align="center">Library size standard deviation (log scale)
parameter. Used to draw library size from a log-normal
distribution.</td>
<td align="center">A single value &gt;=0 indicating the standard
deviation of library size on a log scale.</td>
</tr>
<tr class="even">
<td align="center"><code>customLibSizes</code></td>
<td align="center">-</td>
<td align="center">Optional user-provided library sizes on the original
(non-log) scale. If provided, library sizes will be drawn from this
vector (with replacement) rather than simulated from a log-normal
distribution.</td>
<td align="center">A numeric vector of positive values representing
library sizes on the original scale. If <code>NULL</code> (default),
library sizes are drawn from the specified log-normal distribution using
<code>logLibMean</code> and <code>logLibSD</code>.</td>
</tr>
<tr class="odd">
<td align="center"><code>logLibFacVar</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>α</mi><mi>d</mi></msub><annotation encoding="application/x-tex">\alpha_d</annotation></semantics></math></td>
<td align="center">Variance used for drawing sample-level multiplicative
factors which give different library size distributions for each sample.
Larger values result in larger variation in library size distributions
by sample.</td>
<td align="center">A single value &gt;=0 with 0 indicating no difference
in the library size distributions by sample.</td>
</tr>
<tr class="even">
<td align="center"><code>exprsMean</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mi>g</mi></msub><annotation encoding="application/x-tex">\mu_g</annotation></semantics></math></td>
<td align="center">Gene-specific mean expression value representing the
average expression of each gene in the data set. If a named vector is
given, these names will be used as the gene names in the simulated
<code>SingleCellExperiment</code> object.</td>
<td align="center">Vector of numeric values &gt;=0 with length equal to
desired number of genes in the simulated data where each value indicates
the average expression for a single gene.</td>
</tr>
<tr class="odd">
<td align="center"><code>dispersion</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ϕ</mi><mi>g</mi></msub><annotation encoding="application/x-tex">\phi_g</annotation></semantics></math></td>
<td align="center">Gene-specific dispersion value representing the
variation in expression for each gene in the data set.</td>
<td align="center">Vector of numeric values &gt;0 with length equal to
desired number of genes in the simulated data where each value indicates
the dispersion for a single gene.</td>
</tr>
<tr class="even">
<td align="center"><code>sampleFacVarMean</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mi>b</mi></msub><annotation encoding="application/x-tex">\mu_b</annotation></semantics></math></td>
<td align="center">Mean used for drawing variance (log-scale) of
sample-level multiplicative factors. Larger values result in more
between-sample variation.</td>
<td align="center">A single numeric value.</td>
</tr>
<tr class="odd">
<td align="center"><code>sampleFacVarSD</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>σ</mi><mi>b</mi></msub><annotation encoding="application/x-tex">\sigma_b</annotation></semantics></math></td>
<td align="center">Standard deviation used for drawing variance
(log-scale) of sample-level multiplicative factors. Larger values result
in more variation in amount of between-sample variation across different
genes.</td>
<td align="center">A single numeric value &gt;=0.</td>
</tr>
<tr class="even">
<td align="center"><code>subjectVarMean</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>μ</mi><mi>a</mi></msub><annotation encoding="application/x-tex">\mu_a</annotation></semantics></math></td>
<td align="center">Mean used for drawing variance (log-scale) of
subject-level multiplicative factors. Larger values result in larger
between-subject variation.</td>
<td align="center">A single numeric value.</td>
</tr>
<tr class="odd">
<td align="center"><code>subjectFacVarSD</code></td>
<td align="center"><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>σ</mi><mi>a</mi></msub><annotation encoding="application/x-tex">\sigma_a</annotation></semantics></math></td>
<td align="center">Standard deviation used for drawing variance
(log-scale) of subject-level multiplicative factors. Larger values
result in more variation in amount of between-subject variation across
different genes.</td>
<td align="center">A single numeric value &gt;=0.</td>
</tr>
<tr class="even">
<td align="center"><code>propDE</code></td>
<td align="center">-</td>
<td align="center">Proportion of genes differentially expressed between
time points/groups.</td>
<td align="center">Numeric value between 0 and 1.</td>
</tr>
<tr class="odd">
<td align="center"><code>deLog2FC</code></td>
<td align="center">-</td>
<td align="center">Fold change values used for differentially expressed
genes.</td>
<td align="center">A single numeric value &gt;=0 indicating the positive
log-fold change for differentially expressed genes (assigned positive or
negative at random); a vector of values (positive and negative) to draw
DE
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change values from; a named list of
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change values for all genes in each condition relative to the
reference condition (e.g., <code>time0</code>, <code>group0</code>,
<code>time0_group0</code>). The reference condition should not be
included in the list.</td>
</tr>
</tbody>
</table>
<p>Individual parameter values can be extracted using the
<code>getRescueSimParam</code> function.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Extract nTimepoints parameter</span></span>
<span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="st">"nTimepoints"</span><span class="op">)</span></span>
<span><span class="co">#&gt; numeric(0)</span></span></code></pre></div>
<div class="section level4">
<h4 id="manually-adding-parameter-values">Manually adding parameter values<a class="anchor" aria-label="anchor" href="#manually-adding-parameter-values"></a>
</h4>
<p>Parameter values can be added manually when initially creating the
parameter object by using the parameter name(s) as an argument in the
<code>RescueSimParams</code> function followed by the desired
value(s).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create parameter object with nTimepoints and nSubjsPerGroup pre-set</span></span>
<span><span class="va">myParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RescueSimParams.html">RescueSimParams</a></span><span class="op">(</span>nTimepoints <span class="op">=</span> <span class="fl">2</span>, nSubjsPerGroup <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="st">"nTimepoints"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2</span></span>
<span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="st">"nSubjsPerGroup"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 5</span></span></code></pre></div>
<p>After a parameter object is created, the values can be updated using
<code>updateRescueSimParams</code> with the parameter object and then a
list of parameters and associated values.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Update nTimepoints and nSubjsPerGroup</span></span>
<span><span class="va">myParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span></span>
<span>  paramObj <span class="op">=</span> <span class="va">myParams</span>,</span>
<span>  paramValues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    nTimepoints <span class="op">=</span> <span class="fl">3</span>,</span>
<span>    nSubjsPerGroup <span class="op">=</span> <span class="fl">10</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="st">"nTimepoints"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3</span></span>
<span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="st">"nSubjsPerGroup"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 10</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="estimating-parameter-values-from-empirical-data">Estimating parameter values from empirical data<a class="anchor" aria-label="anchor" href="#estimating-parameter-values-from-empirical-data"></a>
</h4>
<p>Parameters can be estimated from empirical data sets using the
<code>estRescueSimParams</code> function. Users provide a data set in
the form of a <code>SingleCellExperiment</code> object with a counts
matrix in the <code>counts</code> slot and sample, subject, time point,
and group meta data (where applicable) in <code>colData</code>. The
table below outlines other function arguments.</p>
<table class="table">
<colgroup>
<col width="7%">
<col width="89%">
<col width="2%">
</colgroup>
<thead><tr class="header">
<th align="center">Argument</th>
<th>Description</th>
<th align="center">Default</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><code>sce</code></td>
<td>
<code>SingleCellExperiment</code> object containing empirical data
with a counts matrix in the <code>counts</code> slot.</td>
<td align="center">-</td>
</tr>
<tr class="even">
<td align="center"><code>paramObj</code></td>
<td>
<code>RescueSimParams</code> object with empty slots for parameters
which need to be estimated. If <code>NULL</code>, all parameter values
will be estimated.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="odd">
<td align="center"><code>sampleVariable</code></td>
<td>String denoting name of sample identifier variable in the
<code>colData</code> of <code>SingleCellExperiment</code> object. If
<code>NULL</code>, data is assumed to contain only one sample and
parameters <code>sampleFacVarMean</code> and <code>sampleFacVarSD</code>
parameters cannot be estimated.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="even">
<td align="center"><code>subjectVariable</code></td>
<td>String denoting name of subject identifier variable in the
<code>colData</code> of <code>SingleCellExperiment</code> object. If
<code>NULL</code>, the data is assumed to contain only one subject and
parameters <code>nSubjsPerGroup</code>, <code>subjectFacVarMean</code>,
and <code>subjectFacVarSD</code> parameters cannot be estimated.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="odd">
<td align="center"><code>timepointVariable</code></td>
<td>String denoting name of time point identifier variable in the
<code>colData</code> of <code>SingleCellExperiment</code> object. If
<code>NULL</code>, the number of time points for the simulated data is
set to 1.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="even">
<td align="center"><code>groupVariable</code></td>
<td>String denoting name of group identifier variable in the
<code>colData</code> of <code>SingleCellExperiment</code> object. If
<code>NULL</code>, a single group design is assumed.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="odd">
<td align="center"><code>nonDEGs</code></td>
<td>Vector containing names or row indices of genes to be used to
estimate sample and subject factor parameters. Using genes that are
differentially expressed across time points may lead to inaccurate
estimates.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="even">
<td align="center"><code>cellParamsByCondition</code></td>
<td>Logical value indicating whether <code>maxCellsPerSamp</code> and
<code>minCellsPerSamp</code> should be estimated for each condition
(group/time point) or overall.</td>
<td align="center"><code>FALSE</code></td>
</tr>
</tbody>
</table>
<p>The below example shows how we can simulate data set using the
subsetted recruited macrophage data. We estimate all of the parameters
from the data and the empirical data has a single group, so providing a
<code>paramObj</code> or<code>groupVariable</code> is unnecessary. Using
genes that are differentially expressed across time points may lead to
inaccurate estimates to inaccurate estimates of subject and sample
variance parameters (<code>sampleFacVarMean</code>,
<code>sampleFacVarSD</code>, <code>subjectFacVarMean</code>, and
<code>subjectFacVarSD</code>). The recruited macrophage data set only
includes genes that appeared invariant across time points. In data sets
where differential expression is believed to be present, a rough list
invariant genes can be identified using differential expression methods
such as <code>edgeR</code> or <code>DESeq2</code> and the invariant
genes can be passed to the <code>nonDEGs</code> argument.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Estimate all parameters from the data</span></span>
<span><span class="va">myParams_estimated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estRescueSimParams.html">estRescueSimParams</a></span><span class="op">(</span></span>
<span>  sce <span class="op">=</span> <span class="va">RecAM_sce_small</span>, paramObj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sampleVariable <span class="op">=</span> <span class="st">"sampleID"</span>,</span>
<span>  subjectVariable <span class="op">=</span> <span class="st">"subjectID"</span>,</span>
<span>  timepointVariable <span class="op">=</span> <span class="st">"time"</span>,</span>
<span>  groupVariable <span class="op">=</span> <span class="cn">NULL</span>, nonDEGs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellParamsByCondition <span class="op">=</span> <span class="cn">F</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">myParams_estimated</span></span>
<span><span class="co">#&gt; An object of class "RescueSimParams"</span></span>
<span><span class="co">#&gt; Slot "nTimepoints":</span></span>
<span><span class="co">#&gt; [1] 2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "twoGroupDesign":</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "nSubjsPerGroup":</span></span>
<span><span class="co">#&gt; [1] 5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "maxCellsPerSamp":</span></span>
<span><span class="co">#&gt; [1] 136</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "minCellsPerSamp":</span></span>
<span><span class="co">#&gt; [1] 67</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibMean":</span></span>
<span><span class="co">#&gt; [1] 3.178947</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibSD":</span></span>
<span><span class="co">#&gt; [1] 0.522717</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibFacVar":</span></span>
<span><span class="co">#&gt; [1] 0.004475137</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "customLibSizes":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "exprsMean":</span></span>
<span><span class="co">#&gt;       NOC2L      SCNN1D  AL391244.3  AL645728.1  AL691432.2        NADK </span></span>
<span><span class="co">#&gt; 0.572778264 0.008741053 0.009527545 0.071148282 0.072171575 0.980568794 </span></span>
<span><span class="co">#&gt;        GNB1      LRRC47        DFFB    C1orf174      RNF207   CAMTA1-DT </span></span>
<span><span class="co">#&gt; 4.626133728 0.388941905 0.093361548 0.494518440 0.003207091 0.001810623 </span></span>
<span><span class="co">#&gt;      SLC2A5  AL928921.2     TMEM201  PIK3CD-AS2    CTNNBIP1      UBIAD1 </span></span>
<span><span class="co">#&gt; 0.015460749 0.020178445 0.052752156 0.008627416 0.535365356 0.319103416 </span></span>
<span><span class="co">#&gt;        MFN2    SLC25A34      UQCRHL      ZBTB17    ARHGEF19      FBXO42 </span></span>
<span><span class="co">#&gt; 0.510133962 0.039608167 0.030590020 0.115291183 0.015493547 0.599248311 </span></span>
<span><span class="co">#&gt;   ARHGEF10L  AL035413.1        EMC1       CAPZB       TMCO4       PINK1 </span></span>
<span><span class="co">#&gt; 1.497391194 0.011564381 0.443020738 4.303634967 0.233431917 0.579322545 </span></span>
<span><span class="co">#&gt;      HNRNPR      ZNF436         ID3        ELOA      LYPLA2       FUCA1 </span></span>
<span><span class="co">#&gt; 3.324729001 0.049855385 1.503006201 0.722096065 0.623905753 2.399000540 </span></span>
<span><span class="co">#&gt;      SRSF10       SRRM1         RHD     LDLRAP1 </span></span>
<span><span class="co">#&gt; 2.088398063 5.044819599 0.004161722 0.134532354 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "dispersion":</span></span>
<span><span class="co">#&gt;        NOC2L       SCNN1D   AL391244.3   AL645728.1   AL691432.2         NADK </span></span>
<span><span class="co">#&gt; 9.765625e-05 1.024000e+02 9.765625e-05 1.276298e+01 9.765625e-05 1.835820e-01 </span></span>
<span><span class="co">#&gt;         GNB1       LRRC47         DFFB     C1orf174       RNF207    CAMTA1-DT </span></span>
<span><span class="co">#&gt; 1.047582e-01 9.765625e-05 2.531921e+00 9.765625e-05 4.125051e+01 9.765625e-05 </span></span>
<span><span class="co">#&gt;       SLC2A5   AL928921.2      TMEM201   PIK3CD-AS2     CTNNBIP1       UBIAD1 </span></span>
<span><span class="co">#&gt; 9.765625e-05 9.765625e-05 5.431577e+00 1.024000e+02 9.765625e-05 9.765625e-05 </span></span>
<span><span class="co">#&gt;         MFN2     SLC25A34       UQCRHL       ZBTB17     ARHGEF19       FBXO42 </span></span>
<span><span class="co">#&gt; 3.701863e-01 1.611721e+01 9.765625e-05 9.765625e-05 9.765625e-05 2.641163e-01 </span></span>
<span><span class="co">#&gt;    ARHGEF10L   AL035413.1         EMC1        CAPZB        TMCO4        PINK1 </span></span>
<span><span class="co">#&gt; 9.765625e-05 9.765625e-05 9.765625e-05 9.373476e-02 5.244223e-01 2.413057e-01 </span></span>
<span><span class="co">#&gt;       HNRNPR       ZNF436          ID3         ELOA       LYPLA2        FUCA1 </span></span>
<span><span class="co">#&gt; 8.447181e-02 9.765625e-05 2.564846e+00 9.765625e-05 9.765625e-05 2.701521e-01 </span></span>
<span><span class="co">#&gt;       SRSF10        SRRM1          RHD      LDLRAP1 </span></span>
<span><span class="co">#&gt; 1.251372e-01 1.211732e-01 9.765625e-05 2.088006e-01 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "sampleFacVarMean":</span></span>
<span><span class="co">#&gt; [1] -4.105898</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "sampleFacVarSD":</span></span>
<span><span class="co">#&gt; [1] 0.6932764</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "subjectFacVarMean":</span></span>
<span><span class="co">#&gt; [1] -3.853583</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "subjectFacVarSD":</span></span>
<span><span class="co">#&gt; [1] 0.991917</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "propDE":</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "deLog2FC":</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>If you want to manually set some parameters before estimating others,
a <code>RescueSimParams</code> object with blank slots for the
parameters you would like estimated can be provided. Any parameters that
already have values in the <code>RescueSimParams</code> object will not
be estimated. This is particularly useful if using single-sample or
single-time point empirical data. Below is an example of how data can be
simulated using single-time point empirical data after manually setting
some parameters.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subset the data to a single time point</span></span>
<span><span class="va">RecAM_sce_single_time</span> <span class="op">&lt;-</span> <span class="va">RecAM_sce_small</span><span class="op">[</span>, <span class="va">RecAM_sce_small</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="st">"time0"</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Create a params object with necessary parameters filled</span></span>
<span><span class="va">myParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RescueSimParams.html">RescueSimParams</a></span><span class="op">(</span>subjectFacVarMean <span class="op">=</span> <span class="op">-</span><span class="fl">4</span>, subjectFacVarSD <span class="op">=</span> <span class="fl">1</span>, nSubjsPerGroup <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Estimate remaining parameters from the data</span></span>
<span><span class="va">myParams_estimated</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estRescueSimParams.html">estRescueSimParams</a></span><span class="op">(</span></span>
<span>  sce <span class="op">=</span> <span class="va">RecAM_sce_single_time</span>,</span>
<span>  paramObj <span class="op">=</span> <span class="va">myParams</span>,</span>
<span>  sampleVariable <span class="op">=</span> <span class="st">"sampleID"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>All of the parameters are estimated from the data except for the
parameters that we specified when creating the
<code>RescueSimParams</code> object.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myParams_estimated</span></span>
<span><span class="co">#&gt; An object of class "RescueSimParams"</span></span>
<span><span class="co">#&gt; Slot "nTimepoints":</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "twoGroupDesign":</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "nSubjsPerGroup":</span></span>
<span><span class="co">#&gt; [1] 5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "maxCellsPerSamp":</span></span>
<span><span class="co">#&gt; [1] 138</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "minCellsPerSamp":</span></span>
<span><span class="co">#&gt; [1] 74</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibMean":</span></span>
<span><span class="co">#&gt; [1] 3.121858</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibSD":</span></span>
<span><span class="co">#&gt; [1] 0.5599401</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "logLibFacVar":</span></span>
<span><span class="co">#&gt; [1] 0.004115623</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "customLibSizes":</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "exprsMean":</span></span>
<span><span class="co">#&gt;       NOC2L      SCNN1D  AL391244.3  AL645728.1  AL691432.2        NADK </span></span>
<span><span class="co">#&gt; 0.686218318 0.006215246 0.007633068 0.070614455 0.100593001 0.792710639 </span></span>
<span><span class="co">#&gt;        GNB1      LRRC47        DFFB    C1orf174      RNF207   CAMTA1-DT </span></span>
<span><span class="co">#&gt; 4.288298440 0.383281997 0.108778810 0.488922467 0.001990346 0.001987183 </span></span>
<span><span class="co">#&gt;      SLC2A5  AL928921.2     TMEM201  PIK3CD-AS2    CTNNBIP1      UBIAD1 </span></span>
<span><span class="co">#&gt; 0.014821378 0.011496427 0.043983228 0.004346802 0.485034231 0.240097930 </span></span>
<span><span class="co">#&gt;        MFN2    SLC25A34      UQCRHL      ZBTB17    ARHGEF19      FBXO42 </span></span>
<span><span class="co">#&gt; 0.415416269 0.054199585 0.051226562 0.111460850 0.017990270 0.610301394 </span></span>
<span><span class="co">#&gt;   ARHGEF10L  AL035413.1        EMC1       CAPZB       TMCO4       PINK1 </span></span>
<span><span class="co">#&gt; 1.180161571 0.014261528 0.441152623 4.349938165 0.259431948 0.580329663 </span></span>
<span><span class="co">#&gt;      HNRNPR      ZNF436         ID3        ELOA      LYPLA2       FUCA1 </span></span>
<span><span class="co">#&gt; 2.831232806 0.071010670 1.289427830 0.662611225 0.736945307 2.088311882 </span></span>
<span><span class="co">#&gt;      SRSF10       SRRM1         RHD     LDLRAP1 </span></span>
<span><span class="co">#&gt; 2.023526287 5.106106212 0.006105399 0.128508501 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "dispersion":</span></span>
<span><span class="co">#&gt;        NOC2L       SCNN1D   AL391244.3   AL645728.1   AL691432.2         NADK </span></span>
<span><span class="co">#&gt; 1.146676e-01 1.024000e+02 9.765625e-05 1.320901e+01 9.765625e-05 2.842339e-01 </span></span>
<span><span class="co">#&gt;         GNB1       LRRC47         DFFB     C1orf174       RNF207    CAMTA1-DT </span></span>
<span><span class="co">#&gt; 1.295745e-01 9.765625e-05 2.178611e+00 9.765625e-05 9.765625e-05 9.765625e-05 </span></span>
<span><span class="co">#&gt;       SLC2A5   AL928921.2      TMEM201   PIK3CD-AS2     CTNNBIP1       UBIAD1 </span></span>
<span><span class="co">#&gt; 9.765625e-05 9.765625e-05 6.837777e+00 1.024000e+02 9.765625e-05 9.765625e-05 </span></span>
<span><span class="co">#&gt;         MFN2     SLC25A34       UQCRHL       ZBTB17     ARHGEF19       FBXO42 </span></span>
<span><span class="co">#&gt; 3.073198e-01 1.592748e+01 9.765625e-05 9.765625e-05 9.765625e-05 3.012721e-01 </span></span>
<span><span class="co">#&gt;    ARHGEF10L   AL035413.1         EMC1        CAPZB        TMCO4        PINK1 </span></span>
<span><span class="co">#&gt; 9.765625e-05 9.765625e-05 9.765625e-05 1.343976e-01 4.026675e-01 2.689336e-01 </span></span>
<span><span class="co">#&gt;       HNRNPR       ZNF436          ID3         ELOA       LYPLA2        FUCA1 </span></span>
<span><span class="co">#&gt; 8.084418e-02 9.765625e-05 2.418655e+00 9.765625e-05 9.765625e-05 3.516693e-01 </span></span>
<span><span class="co">#&gt;       SRSF10        SRRM1          RHD      LDLRAP1 </span></span>
<span><span class="co">#&gt; 1.393887e-01 1.060177e-01 9.765625e-05 4.842573e-02 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "sampleFacVarMean":</span></span>
<span><span class="co">#&gt; [1] -3.846468</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "sampleFacVarSD":</span></span>
<span><span class="co">#&gt; [1] 0.7435008</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "subjectFacVarMean":</span></span>
<span><span class="co">#&gt; [1] -4</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "subjectFacVarSD":</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "propDE":</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slot "deLog2FC":</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>If <code>propDE</code> and <code>deLog2FC</code> are not manually
set, they are both set to 0 (no differential expression).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams_estimated</span>, <span class="st">"propDE"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams_estimated</span>, <span class="st">"deLog2FC"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="simulating-data">Simulating Data<a class="anchor" aria-label="anchor" href="#simulating-data"></a>
</h2>
<p>After setting simulation parameters, the <code>simRescueData</code>
function can be used to simulate a data set. This function takes as an
argument a <code>RescueSimParams</code> object. It returns a
<code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Simulate data</span></span>
<span><span class="va">mySim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_estimated</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Look at class of simulated data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">mySim</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SingleCellExperiment"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "SingleCellExperiment"</span></span></code></pre></div>
<p>A matrix of counts is held the <code>counts</code> slot. The rownames
(genes) take on the names of the <code>exprsMean</code> vector if
provided. If <code>exprsMean</code> was estimated from empirical data,
these are the gene names from the empirical data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">mySim</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;            cell_1 cell_2 cell_3 cell_4 cell_5</span></span>
<span><span class="co">#&gt; NOC2L           0      0      2      0      1</span></span>
<span><span class="co">#&gt; SCNN1D          0      0      0      0      0</span></span>
<span><span class="co">#&gt; AL391244.3      0      0      0      0      0</span></span>
<span><span class="co">#&gt; AL645728.1      0      0      0      0      0</span></span>
<span><span class="co">#&gt; AL691432.2      0      0      0      0      0</span></span></code></pre></div>
<p>Cell level meta data is held in the object <code>colData</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">mySim</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">#&gt;           sampleID   subjectID        time       group</span></span>
<span><span class="co">#&gt;        &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; cell_1     sample1    subject1       time0      group0</span></span>
<span><span class="co">#&gt; cell_2     sample1    subject1       time0      group0</span></span>
<span><span class="co">#&gt; cell_3     sample1    subject1       time0      group0</span></span>
<span><span class="co">#&gt; cell_4     sample1    subject1       time0      group0</span></span>
<span><span class="co">#&gt; cell_5     sample1    subject1       time0      group0</span></span>
<span><span class="co">#&gt; cell_6     sample1    subject1       time0      group0</span></span></code></pre></div>
<p>Log-fold change values for each gene are held in the object
<code>rowData</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">mySim</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 0 columns</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="adjusting-specific-parameter-values">Adjusting Specific Parameter Values<a class="anchor" aria-label="anchor" href="#adjusting-specific-parameter-values"></a>
</h2>
<p>Now we’ll give more detail on the effect of adjusting some of the
parameter values. We first estimate parameters using the unfiltered
recruited macrophage data. We use the resulting object throughout the
remainder of the document.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Estimate parameters using full RecAM_sce object</span></span>
<span><span class="va">myParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estRescueSimParams.html">estRescueSimParams</a></span><span class="op">(</span><span class="va">RecAM_sce</span>,</span>
<span>  sampleVariable <span class="op">=</span> <span class="st">"sampleID"</span>,</span>
<span>  subjectVariable <span class="op">=</span> <span class="st">"subjectID"</span>,</span>
<span>  timepointVariable <span class="op">=</span> <span class="st">"time"</span>,</span>
<span>  groupVariable <span class="op">=</span> <span class="cn">NULL</span>, nonDEGs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cellParamsByCondition <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="sample-size-adjustments">Sample Size Adjustments<a class="anchor" aria-label="anchor" href="#sample-size-adjustments"></a>
</h3>
<p>A question likely to arise in repeated measures scRNA-seq analysis
and study planning is how the sample size will impact results. In these
experiments, there are different aspects of sample size, all of which
can be controlled in using the <code>rescueSim</code> simulation
method:</p>
<ul>
<li><p>Number of time points (samples) per subject
(<code>nTimepoints</code>)</p></li>
<li><p>Number of subjects (<code>nSubjsPerGroup</code>)</p></li>
<li><p>Number of cells per sample (<code>maxCellsPerSamp</code>,
<code>maxCellsPerSamp</code>)</p></li>
</ul>
<p>A situation that sometimes arises in scRNA-seq experiments is one in
which the number of cells of a particular cell type in a sample is
impacted by a group variable (i.e. treatment or control) or a time
variable (ex. pre or post-treatment). Thus, we allow the option to
specify or estimate one <code>maxCellsPerSamp</code>and
<code>maxCellsPerSamp</code> parameter to be used for all samples, one
<code>maxCellsPerSamp</code>and <code>maxCellsPerSamp</code> parameter
for each condition (group/time), or one <code>maxCellsPerSamp</code>and
<code>maxCellsPerSamp</code> parameter for each individual sample. Below
is an example of estimating or manually specifying
<code>maxCellsPerSamp</code>and <code>maxCellsPerSamp</code> parameters
for the two time points.</p>
<p>A user can estimate individual parameters for each condition by using
<code>cellParamsByCondition=TRUE</code> as a function argument.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set cell parameter values to numeric 0 so we can re-estimate</span></span>
<span><span class="va">myParams_cellsByCondition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span></span>
<span>  <span class="va">myParams</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    maxCellsPerSamp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>    minCellsPerSamp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Estimate with cellParamsByCondition = T</span></span>
<span><span class="va">myParams_cellsByCondition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/estRescueSimParams.html">estRescueSimParams</a></span><span class="op">(</span><span class="va">RecAM_sce</span>,</span>
<span>  <span class="va">myParams_cellsByCondition</span>,</span>
<span>  sampleVariable <span class="op">=</span> <span class="st">"sampleID"</span>,</span>
<span>  subjectVariable <span class="op">=</span> <span class="st">"subjectID"</span>,</span>
<span>  timepointVariable <span class="op">=</span> <span class="st">"time"</span>,</span>
<span>  cellParamsByCondition <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Now cell parameters are estimated separately by time point</span></span>
<span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams_cellsByCondition</span>, <span class="st">"maxCellsPerSamp"</span><span class="op">)</span></span>
<span><span class="co">#&gt; time0 time1 </span></span>
<span><span class="co">#&gt;   138   104</span></span>
<span><span class="fu"><a href="../reference/getRescueSimParam.html">getRescueSimParam</a></span><span class="op">(</span><span class="va">myParams_cellsByCondition</span>, <span class="st">"minCellsPerSamp"</span><span class="op">)</span></span>
<span><span class="co">#&gt; time0 time1 </span></span>
<span><span class="co">#&gt;    74    67</span></span></code></pre></div>
<p>Alternatively, a user can manually set the parameters with vector
values equal to the number of conditions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Manually setting parameters</span></span>
<span><span class="va">myParams_cellsByCondition</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span></span>
<span>  <span class="va">myParams</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    maxCellsPerSamp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">500</span><span class="op">)</span>,</span>
<span>    minCellsPerSamp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">250</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>When we simulate data, we can see that the number of cells for time0
samples are all between 50 and 100 while the number of cells for time1
samples are between 250 and 500.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Simulate data</span></span>
<span><span class="va">simDat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_cellsByCondition</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Look at number of cells per sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">simDat</span><span class="op">$</span><span class="va">sampleID</span>, <span class="va">simDat</span><span class="op">$</span><span class="va">time</span><span class="op">)</span></span>
<span><span class="co">#&gt;           </span></span>
<span><span class="co">#&gt;            time0 time1</span></span>
<span><span class="co">#&gt;   sample1     69     0</span></span>
<span><span class="co">#&gt;   sample10     0   368</span></span>
<span><span class="co">#&gt;   sample2     75     0</span></span>
<span><span class="co">#&gt;   sample3     72     0</span></span>
<span><span class="co">#&gt;   sample4     72     0</span></span>
<span><span class="co">#&gt;   sample5     77     0</span></span>
<span><span class="co">#&gt;   sample6      0   361</span></span>
<span><span class="co">#&gt;   sample7      0   310</span></span>
<span><span class="co">#&gt;   sample8      0   355</span></span>
<span><span class="co">#&gt;   sample9      0   257</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="library-size-adjustments">Library Size Adjustments<a class="anchor" aria-label="anchor" href="#library-size-adjustments"></a>
</h3>
<p>The library sizes, or number of counts per cell, can be adjusted in
two ways. By default, library sizes are drawn from a log-normal
distribution where <code>logLibMean</code> the <code>logLibSD</code>
control the overall mean and variability of library sizes across cells.
Alternatively, users can supply specific library sizes through the
<code>customLibSizes</code> parameter from which library sizes are drawn
directly rather than simulated from a distribution. In either case,
<code>logLibFacVar</code> introduces variation in library size
distributions between samples, mimicking technical effects such as batch
variation.</p>
<p>The <code>logLibMean</code> parameter controls the overall average
library size (log scale). Below is an example of using a relatively
large vs. small value for <code>logLibMean</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create two parameter objects, one with large logLibMean, one with small logLibMean</span></span>
<span><span class="va">myParams_bigLogLibMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logLibMean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">5000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">myParams_smallLogLibMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logLibMean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">500</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Simulate data with both scenarios</span></span>
<span><span class="va">simDat_bigLogLibMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_bigLogLibMean</span><span class="op">)</span></span>
<span><span class="va">simDat_smallLogLibMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_smallLogLibMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate library sizes by summing up the columns of the count data</span></span>
<span><span class="va">bigLibSizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  libSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">simDat_bigLogLibMean</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"logLibMean = log(5,000)"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">smallLibSizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  libSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">simDat_smallLogLibMean</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"logLibMean = log(500)"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">libSizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">bigLibSizes</span>, <span class="va">smallLibSizes</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Output boxplot comparing library sizes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">libSizes</span><span class="op">)</span> <span class="op">~</span> <span class="va">type</span>, data <span class="op">=</span> <span class="va">libSizes</span>, xlab <span class="op">=</span> <span class="cn">NULL</span>, ylab <span class="op">=</span> <span class="st">"log(Library Size)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>The <code>logLibSD</code> parameter controls the amount of variation
in library size across cells. Below is an example of using a relatively
large vs. small value for <code>logLibSD</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create two parameter objects, one with large logLibSD, one with small logLibSD</span></span>
<span><span class="va">myParams_bigLogLibSD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logLibSD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">myParams_smallLogLibSD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logLibSD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1.25</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Simulate data with both scenarios</span></span>
<span><span class="va">simDat_bigLogLibSD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_bigLogLibSD</span><span class="op">)</span></span>
<span><span class="va">simDat_smallLogLibSD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_smallLogLibSD</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate library sizes by summing up the columns of the count data</span></span>
<span><span class="va">bigLibSizeSD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  libSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">simDat_bigLogLibSD</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"logLibSD = log(2)"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">smallLibSizeSD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  libSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">simDat_smallLogLibSD</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"logLibSD = log(1.25)"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">libSizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">bigLibSizeSD</span>, <span class="va">smallLibSizeSD</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Output boxplot comparing library sizes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">libSizes</span><span class="op">)</span> <span class="op">~</span> <span class="va">type</span>, data <span class="op">=</span> <span class="va">libSizes</span>, xlab <span class="op">=</span> <span class="cn">NULL</span>, ylab <span class="op">=</span> <span class="st">"log(Library Size)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>Alternatively, users can provide custom library sizes via the
<code>customLibSizes</code> parameter. When specified, these values are
used directly (sampled with replacement) instead of drawing library
sizes from a log-normal distribution. This can be useful in cases where
the log-normal distribution does not reflect the empirical distribution
of library sizes.</p>
<p>Below we illustrate how you could create custom library sizes based
of an empirical distribution by identifying the sample in the data set
with a mean library size closest to the overall mean across samples, and
use that sample’s library sizes for simulation.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Calculate library sizes for each cell</span></span>
<span><span class="va">empLibSizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  libSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  sample <span class="op">=</span> <span class="va">RecAM_sce</span><span class="op">$</span><span class="va">sampleID</span>,</span>
<span>  type<span class="op">=</span><span class="st">"Empirical"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate mean library size for each sample</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">empLibSizes</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="va">lib_size_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">samples</span>, <span class="kw">function</span><span class="op">(</span><span class="va">s</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">empLibSizes</span><span class="op">$</span><span class="va">libSizes</span><span class="op">[</span><span class="va">empLibSizes</span><span class="op">$</span><span class="va">sample</span> <span class="op">==</span> <span class="va">s</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Find the sample closest to the overall mean</span></span>
<span><span class="va">overall_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">lib_size_means</span><span class="op">)</span></span>
<span><span class="va">idx_min</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">lib_size_means</span> <span class="op">-</span> <span class="va">overall_mean</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">samp_min</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span><span class="va">idx_min</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Get custom library sizes from that sample</span></span>
<span><span class="va">customLibSizes</span> <span class="op">&lt;-</span> <span class="va">empLibSizes</span><span class="op">$</span><span class="va">libSizes</span><span class="op">[</span><span class="va">empLibSizes</span><span class="op">$</span><span class="va">sample</span> <span class="op">==</span> <span class="va">samp_min</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Update parameters and simulate</span></span>
<span><span class="va">myParams_customLib</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>customLibSizes <span class="op">=</span> <span class="va">customLibSizes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">simDat_customLib</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_customLib</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Compare empirical and simulated library sizes</span></span>
<span></span>
<span><span class="va">simLibSizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  libSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">simDat_customLib</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  sample<span class="op">=</span><span class="va">simDat_customLib</span><span class="op">$</span><span class="va">sampleID</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"Simulated (customLibSizes)"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">libSizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">empLibSizes</span>, <span class="va">simLibSizes</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Boxplot comparing library sizes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">libSizes</span><span class="op">)</span> <span class="op">~</span> <span class="va">type</span>,</span>
<span>        data<span class="op">=</span><span class="va">libSizes</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"log(Library Size)"</span>, xlab <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>In multi-sample scRNA-Seq data, there are often differences in the
distributions of library size across samples caused by technical batch
effects since cells from individual samples are sequenced together in a
batch. To account for this, we draw multiplicative factors, centered
around one, which we multiply with <code>logLibMean</code> so that the
mean used to draw the library sizes varies by sample. The parameter
<code>logLibFacVar</code> represents the estimated variance between
these multiplicative factors. A value of 0 means there is no variance in
the library size means, and thus the distribution of library sizes is
approximately the same for each sample. Lager values of
<code>logLibFacVar</code> indicate larger between-sample differences in
library size distribution. Below is an example of using a big and small
value for <code>logLibFacVar</code> and how it affects the library size
distribution by sample.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create two parameter objects, one with large logLibFacVar, one with small logLibFacVar</span></span>
<span><span class="va">myParams_smalllogLibFacVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logLibFacVar <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">myParams_biglogLibFacVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logLibFacVar <span class="op">=</span> <span class="fl">.005</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## Simulate data with both scenarios</span></span>
<span><span class="va">simDat_smalllogLibFacVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_smalllogLibFacVar</span><span class="op">)</span></span>
<span><span class="va">simDat_biglogLibFacVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_biglogLibFacVar</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Calculate library sizes by summing up the columns of the count data</span></span>
<span><span class="va">smalllogLibFacVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  libSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">simDat_smalllogLibFacVar</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"Small logLibFacVar"</span>,</span>
<span>  sample <span class="op">=</span> <span class="va">simDat_smalllogLibFacVar</span><span class="op">$</span><span class="va">sampleID</span></span>
<span><span class="op">)</span></span>
<span><span class="va">biglogLibFacVar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  libSizes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">simDat_biglogLibFacVar</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"Big logLibFacVar"</span>,</span>
<span>  sample <span class="op">=</span> <span class="va">simDat_biglogLibFacVar</span><span class="op">$</span><span class="va">sampleID</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## Output boxplot comparing library sizes by samples for both scenarios</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">libSizes</span><span class="op">)</span> <span class="op">~</span> <span class="va">sample</span>,</span>
<span>  data <span class="op">=</span> <span class="va">smalllogLibFacVar</span>, main <span class="op">=</span> <span class="st">"logLibFacVar = 0"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="cn">NULL</span>, ylab <span class="op">=</span> <span class="st">"log(Library Size)"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">libSizes</span><span class="op">)</span> <span class="op">~</span> <span class="va">sample</span>,</span>
<span>  data <span class="op">=</span> <span class="va">biglogLibFacVar</span>, main <span class="op">=</span> <span class="st">"logLibFacVar = .005"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="cn">NULL</span>, ylab <span class="op">=</span> <span class="st">"log(Library Size)"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-23-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="between-subjectsample-variation">Between Subject/Sample Variation<a class="anchor" aria-label="anchor" href="#between-subjectsample-variation"></a>
</h3>
<p>To simulate between-sample and between-subject variation,
<code>rescueSim</code> uses gene specific sample- and subject- level
multiplicative factors which are multiplied with the mean gene
expression for each gene, with the resulting value being used later in
simulation (see Figure 1, step 1). The multiplicative factors are drawn
from a distribution with a mean of 1. The variance of this distribution
is randomly generated. The parameters <code>sampleFacVarMean</code> and
<code>sampleFacVarSD</code> represent the mean and standard deviation
(log scale) of the variances for the sample-level multiplicative factors
while <code>subjectFacVarMean</code> and <code>subjectFacVarSD</code>
are the parameters for the sampling distribution of subject-level
variances. Because larger variances result in larger between-sample or
-subject variation, increasing the <code>sampleFacVarMean</code> or
<code>subjectFacVarMean</code> also leads to more between-sample or
-subject variation. Increasing the values of <code>sampleFacVarSD</code>
and <code>subjectFacVarSD</code> lead to in more variation in amount of
between-sample or -subject variation across genes.</p>
<p>Typically, we are more interested in adjusting the overall amount of
between-sample or -subject variation, so we will concentrate on the
effect of changing the <code>sampleFacVarMean</code> and
<code>subjectFacVarMean</code> parameters.</p>
<p>Using a low value (-10) for <code>sampleFacVarMean</code> leads to
little between-sample variation, and so we see in the TSNE plot that the
samples don’t cluster together</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set sampleFacVarMean to -10</span></span>
<span><span class="va">myParams_smallsampleFacVarMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sampleFacVarMean <span class="op">=</span> <span class="op">-</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Simulate data</span></span>
<span><span class="va">simDat_smallsampleFacVarMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_smallsampleFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log-normalize counts to run TSNE</span></span>
<span><span class="va">simDat_smallsampleFacVarMean</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_smallsampleFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run and plot TSNE</span></span>
<span><span class="va">simDat_smallsampleFacVarMean</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_smallsampleFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_smallsampleFacVarMean</span>, colour_by <span class="op">=</span> <span class="st">"sampleID"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-24-1.png" width="500"></p>
<p>If we increase the value to -3, the samples cluster very tightly by
sample.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set sampleFacVarMean to -3</span></span>
<span><span class="va">myParams_bigsampleFacVarMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>sampleFacVarMean <span class="op">=</span> <span class="op">-</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Simulate data</span></span>
<span><span class="va">simDat_bigsampleFacVarMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_bigsampleFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log-normalize counts to run TSNE</span></span>
<span><span class="va">simDat_bigsampleFacVarMean</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_bigsampleFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run and plot TSNE</span></span>
<span><span class="va">simDat_bigsampleFacVarMean</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_bigsampleFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_bigsampleFacVarMean</span>, colour_by <span class="op">=</span> <span class="st">"sampleID"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-25-1.png" width="500"></p>
<p>The same is true if we look at the effect of increasing the
<code>subjectFacVarMean</code> on subject-level clustering.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set subjectFacVarMean to -3</span></span>
<span><span class="va">myParams_smallsubjectFacVarMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>subjectFacVarMean <span class="op">=</span> <span class="op">-</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Simulate data</span></span>
<span><span class="va">simDat_smallsubjectFacVarMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_smallsubjectFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log-normalize counts to run TSNE</span></span>
<span><span class="va">simDat_smallsubjectFacVarMean</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_smallsubjectFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run and plot TSNE</span></span>
<span><span class="va">simDat_smallsubjectFacVarMean</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_smallsubjectFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_smallsubjectFacVarMean</span>, colour_by <span class="op">=</span> <span class="st">"subjectID"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-26-1.png" width="500"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Set subjectFacVarMean to -3</span></span>
<span><span class="va">myParams_bigsubjectFacVarMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>subjectFacVarMean <span class="op">=</span> <span class="op">-</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Simulate data</span></span>
<span><span class="va">simDat_bigsubjectFacVarMean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_bigsubjectFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log-normalize counts to run TSNE</span></span>
<span><span class="va">simDat_bigsubjectFacVarMean</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_bigsubjectFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run and plot TSNE</span></span>
<span><span class="va">simDat_bigsubjectFacVarMean</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_bigsubjectFacVarMean</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_bigsubjectFacVarMean</span>, colour_by <span class="op">=</span> <span class="st">"subjectID"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-27-1.png" width="500"></p>
</div>
<div class="section level3">
<h3 id="differential-expression">Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h3>
<p><code>rescueSim</code> allows the user to simulate differential
expression between time points/groups using two parameters which must be
manually set by the user: <code>propDE</code> which controls the
proportion of genes differentially expressed, and <code>deLog2FC</code>
which controls the
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change values of differentially expressed genes. A user can specify
a single positive
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change value for <code>deLog2FC</code>, in which case differential
expression is evenly split between up- and down-regulation in the
proportion of genes specified by <code>propDE</code>. Alternatively, a
vector of positive and/or negative values can be provided, from which
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change values from differentially expressed genes are drawn. In
both of these cases, when simulating multiple time points, the
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change represents the change between the first (<code>time0</code>)
and last time point, with gene expression changes simulated linearly
across intermediate time points. In a two-group, multi-time point
simulation, the
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change represents the difference between <code>time0_group0</code>
and the final time point of <code>group1</code>
(<code>timeN_group1</code>), with no differential expression across time
points within group0 and a linear change across time within
<code>group1</code>.</p>
<p>Alternatively, users can provide a named list of vectors to
<code>deLogFC</code> where each element of the list corresponds to the
fold change values at each time point (or time point-group combination)
relative to the baseline (<code>time0</code>, <code>group0</code>, or
<code>time0_group0</code>). This enables simulation of non-linear or
custom differential expression patterns across time and groups. When
using a list, the <code>propDE</code> parameter is not used and each
vector should have length equal to the number of genes in the data set.
The baseline condition is not included in the list.</p>
<p>We show below using a TSNE plot from the <code>scater</code> package
how differential expression impacts cell clustering.</p>
<p>If we simulate with no differential expression, cells from both time
points cluster together.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Estimate data with no differential expression (default)</span></span>
<span><span class="va">simDat_noDE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log-normalize counts to run TSNE</span></span>
<span><span class="va">simDat_noDE</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_noDE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run and plot TSNE</span></span>
<span><span class="va">simDat_noDE</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_noDE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_noDE</span>, colour_by <span class="op">=</span> <span class="st">"time"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-28-1.png" width="500"></p>
<p>If simulate 20% of genes to be differentially expressed between time
points with a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>±</mo><mn>1</mn></mrow><annotation encoding="application/x-tex">\pm1</annotation></semantics></math>
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change, cells from different time points cluster separately.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Manually changing DE Params</span></span>
<span><span class="va">myParams_de_2timepoints</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  propDE <span class="op">=</span> <span class="fl">.2</span>, deLog2FC <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Estimate data with no differential expression (default)</span></span>
<span><span class="va">simDat_de_2timepoints</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_de_2timepoints</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log-normalize counts to run TSNE</span></span>
<span><span class="va">simDat_de_2timepoints</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_de_2timepoints</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run and plot TSNE</span></span>
<span><span class="va">simDat_de_2timepoints</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_de_2timepoints</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_de_2timepoints</span>, colour_by <span class="op">=</span> <span class="st">"time"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-29-1.png" width="500"></p>
<p>If we simulate three time points, the cells cluster sequentially by
time point since <code>rescueSim</code> simulates linear differential
expression across time.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Manually changing DE Params</span></span>
<span><span class="va">myParams_de_3timepoints</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  propDE <span class="op">=</span> <span class="fl">.2</span>, deLog2FC <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  nTimepoints <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Estimate data with no differential expression (default)</span></span>
<span><span class="va">simDat_de_3timepoints</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_de_3timepoints</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log-normalize counts to run TSNE</span></span>
<span><span class="va">simDat_de_3timepoints</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_de_3timepoints</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run and plot TSNE</span></span>
<span><span class="va">simDat_de_3timepoints</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_de_3timepoints</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_de_3timepoints</span>, colour_by <span class="op">=</span> <span class="st">"time"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-30-1.png" width="500"></p>
<p>If we simulate a 2-group, 2-time point data set, then the cells from
the final time point and last group, which is simulated to be
differentially expressed relative to the other groups/times, cluster
separately from the rest of the cells.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Manually changing DE Params</span></span>
<span><span class="va">myParams_de_2groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  propDE <span class="op">=</span> <span class="fl">.2</span>, deLog2FC <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  twoGroupDesign <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  nTimepoints <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Estimate data with no differential expression (default)</span></span>
<span><span class="va">simDat_de_2groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_de_2groups</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Make a variable combining group and time</span></span>
<span><span class="va">simDat_de_2groups</span><span class="op">$</span><span class="va">group_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">simDat_de_2groups</span><span class="op">$</span><span class="va">group</span>,</span>
<span>  <span class="va">simDat_de_2groups</span><span class="op">$</span><span class="va">time</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">"_"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Log-normalize counts to run TSNE</span></span>
<span><span class="va">simDat_de_2groups</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_de_2groups</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Run and plot TSNE</span></span>
<span><span class="va">simDat_de_2groups</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_de_2groups</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_de_2groups</span>, colour_by <span class="op">=</span> <span class="st">"group_time"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-31-1.png" width="500"></p>
<p>Finally, we can simulate differential expression using manually set
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change values for each time point or condition, allowing for
non-linear changes across time. This is done by providing
<code>deLog2FC</code> as a list of named vectors, where each element of
the list corresponds to a condition (excluding the reference:
<code>time0</code>, <code>group0</code> or <code>time0_group0</code>).
Each vector specifies the
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold change values for each gene in that condition relative to the
reference group.</p>
<p>Below, we show an example of simulating three time points, where we
specify
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold changes for one intermediate time point and larger
log_<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>2</mn><annotation encoding="application/x-tex">2</annotation></semantics></math>
fold changes for the final time point. This results in non-linear gene
expression changes across time.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## For the 2nd time point, simulate +/- 1 log2FC in 20% of genes</span></span>
<span><span class="va">time1_log2FC</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">)</span>, replace<span class="op">=</span><span class="cn">T</span>, prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.1</span>, <span class="fl">.8</span>, <span class="fl">.1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## For final time point, multiply log2FC by 3 to get +/- 3 log2FC in 20% of genes</span></span>
<span><span class="va">time2_log2FC</span> <span class="op">&lt;-</span> <span class="va">time1_log2FC</span><span class="op">*</span><span class="fl">3</span></span>
<span></span>
<span><span class="va">myParams_nonLinearDE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">myParams</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  deLog2FC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>time1<span class="op">=</span><span class="va">time1_log2FC</span>, time2<span class="op">=</span><span class="va">time2_log2FC</span><span class="op">)</span>,</span>
<span>  nTimepoints <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">simDat_nonLinearDE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simRescueData.html">simRescueData</a></span><span class="op">(</span><span class="va">myParams_nonLinearDE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Log-normalize counts and run TSNE</span></span>
<span><span class="va">simDat_nonLinearDE</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">simDat_nonLinearDE</span><span class="op">)</span></span>
<span><span class="va">simDat_nonLinearDE</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">simDat_nonLinearDE</span><span class="op">)</span></span></code></pre></div>
<p>In this case, we can see that the time0 and time1 are much closer
together than to time2.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot TSNE colored by time point to visualize clustering</span></span>
<span><span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">simDat_nonLinearDE</span>, colour_by <span class="op">=</span> <span class="st">"time"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rescueSimVignette_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SingleCellExperiment_1.30.1 SummarizedExperiment_1.38.1</span></span>
<span><span class="co">#&gt;  [3] Biobase_2.68.0              GenomicRanges_1.60.0       </span></span>
<span><span class="co">#&gt;  [5] GenomeInfoDb_1.44.0         IRanges_2.42.0             </span></span>
<span><span class="co">#&gt;  [7] S4Vectors_0.46.0            BiocGenerics_0.54.0        </span></span>
<span><span class="co">#&gt;  [9] generics_0.1.4              MatrixGenerics_1.20.0      </span></span>
<span><span class="co">#&gt; [11] matrixStats_1.5.0           rescueSim_0.99.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1        viridisLite_0.4.2       dplyr_1.1.4            </span></span>
<span><span class="co">#&gt;  [4] vipor_0.4.7             farver_2.1.2            viridis_0.6.5          </span></span>
<span><span class="co">#&gt;  [7] fastmap_1.2.0           bluster_1.18.0          digest_0.6.37          </span></span>
<span><span class="co">#&gt; [10] rsvd_1.0.5              lifecycle_1.0.4         cluster_2.1.8.1        </span></span>
<span><span class="co">#&gt; [13] statmod_1.5.0           magrittr_2.0.3          compiler_4.5.1         </span></span>
<span><span class="co">#&gt; [16] rlang_1.1.6             sass_0.4.10             tools_4.5.1            </span></span>
<span><span class="co">#&gt; [19] igraph_2.1.4            yaml_2.3.10             data.table_1.17.6      </span></span>
<span><span class="co">#&gt; [22] knitr_1.50              labeling_0.4.3          dqrng_0.4.1            </span></span>
<span><span class="co">#&gt; [25] S4Arrays_1.8.1          DelayedArray_0.34.1     plyr_1.8.9             </span></span>
<span><span class="co">#&gt; [28] RColorBrewer_1.1-3      abind_1.4-8             BiocParallel_1.42.1    </span></span>
<span><span class="co">#&gt; [31] Rtsne_0.17              withr_3.0.2             desc_1.4.3             </span></span>
<span><span class="co">#&gt; [34] grid_4.5.1              beachmat_2.24.0         edgeR_4.6.2            </span></span>
<span><span class="co">#&gt; [37] ggplot2_3.5.2           scales_1.4.0            gtools_3.9.5           </span></span>
<span><span class="co">#&gt; [40] MAST_1.33.0             cli_3.6.5               rmarkdown_2.29         </span></span>
<span><span class="co">#&gt; [43] crayon_1.5.3            ragg_1.4.0              metapod_1.16.0         </span></span>
<span><span class="co">#&gt; [46] httr_1.4.7              reshape2_1.4.4          scuttle_1.18.0         </span></span>
<span><span class="co">#&gt; [49] ggbeeswarm_0.7.2        cachem_1.1.0            stringr_1.5.1          </span></span>
<span><span class="co">#&gt; [52] parallel_4.5.1          XVector_0.48.0          vctrs_0.6.5            </span></span>
<span><span class="co">#&gt; [55] Matrix_1.7-3            jsonlite_2.0.0          BiocSingular_1.24.0    </span></span>
<span><span class="co">#&gt; [58] BiocNeighbors_2.2.0     ggrepel_0.9.6           irlba_2.3.5.1          </span></span>
<span><span class="co">#&gt; [61] beeswarm_0.4.0          scater_1.36.0           systemfonts_1.2.3      </span></span>
<span><span class="co">#&gt; [64] locfit_1.5-9.12         limma_3.64.1            jquerylib_0.1.4        </span></span>
<span><span class="co">#&gt; [67] glue_1.8.0              pkgdown_2.1.3           codetools_0.2-20       </span></span>
<span><span class="co">#&gt; [70] stringi_1.8.7           gtable_0.3.6            UCSC.utils_1.4.0       </span></span>
<span><span class="co">#&gt; [73] ScaledMatrix_1.16.0     tibble_3.3.0            pillar_1.10.2          </span></span>
<span><span class="co">#&gt; [76] htmltools_0.5.8.1       GenomeInfoDbData_1.2.14 R6_2.6.1               </span></span>
<span><span class="co">#&gt; [79] textshaping_1.0.1       evaluate_1.0.4          lattice_0.22-7         </span></span>
<span><span class="co">#&gt; [82] backports_1.5.0         scran_1.36.0            bslib_0.9.0            </span></span>
<span><span class="co">#&gt; [85] Rcpp_1.0.14             gridExtra_2.3           SparseArray_1.8.0      </span></span>
<span><span class="co">#&gt; [88] checkmate_2.3.2         xfun_0.52               fs_1.6.6               </span></span>
<span><span class="co">#&gt; [91] pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Elizabeth Wynn.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
