<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Power Analysis with rescueSim • rescueSim</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Power Analysis with rescueSim">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rescueSim</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/rescueSimPowerVignette.html">Power Analysis with rescueSim</a></li>
    <li><a class="dropdown-item" href="../articles/rescueSimVignette.html">Intro to rescueSim</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ewynn610/rescueSim/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Power Analysis with rescueSim</h1>
            
            <h4 data-toc-skip class="date">June 30, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ewynn610/rescueSim/blob/master/vignettes/rescueSimPowerVignette.Rmd" class="external-link"><code>vignettes/rescueSimPowerVignette.Rmd</code></a></small>
      <div class="d-none name"><code>rescueSimPowerVignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>One application of the <code>rescueSim</code> is to evaluate the
power to detect within-cell type differential gene expression in paired
and longitudinal scRNA-sequencing studies, as well as other complex
repeated measures designs. These data are inherently high-dimensional
and hierarchical, with multiple sources of variability, making standard
closed-form power or sample size calculations infeasible. Instead,
simulating data and assessing power across different design scenarios
(e.g., varying numbers of samples or cells) offers a practical and
informative approach to study planning.</p>
<p>This vignette focuses on using <code>rescueSim</code> to perform
power analysis for detecting differential gene expression in
longitudinal or repeated-measures single-cell RNA-seq studies. For an
overview of the <code>rescueSim</code> simulation framework and basic
usage, see the Intro to <code>rescueSim</code> vignette.</p>
</div>
<div class="section level2">
<h2 id="running-a-power-analysis">Running a Power Analysis<a class="anchor" aria-label="anchor" href="#running-a-power-analysis"></a>
</h2>
<p>Power can be evaluated using the <code>runRescueSimPower</code>
function. This function simulates multiple data sets under different
user-defined scenarios, applies a user-supplied differential expression
(DE) method, and calculates power as the proportion of true DE genes
correctly identified at a specified significance threshold. The table
below outlines key function arguments.</p>
<table class="table">
<colgroup>
<col width="6%">
<col width="90%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th align="center">Argument</th>
<th>Description</th>
<th align="center">Default</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><code>baseParams</code></td>
<td>A <code>RescueSimParams</code> object containing the base simulation
parameters. These will be updated according to each scenario.</td>
<td align="center">—</td>
</tr>
<tr class="even">
<td align="center"><code>scenarios</code></td>
<td>A <code>data.frame</code> specifying the scenarios to simulate. Each
row corresponds to one scenario, with columns for parameter values to
update in <code>baseParams</code>. Columns must match fields in
<code>RescueSimParams</code>.</td>
<td align="center">—</td>
</tr>
<tr class="odd">
<td align="center"><code>deFunction</code></td>
<td>A user-supplied function that takes a simulated
<code>SingleCellExperiment</code> object and returns a
<code>data.frame</code> with columns <code>gene</code> and
<code>padj</code>.</td>
<td align="center">—</td>
</tr>
<tr class="even">
<td align="center"><code>nSim</code></td>
<td>Number of simulations to run for each scenario.</td>
<td align="center">1</td>
</tr>
<tr class="odd">
<td align="center"><code>padjThresh</code></td>
<td>Significance threshold (e.g., FDR cutoff) to define true positive DE
calls.</td>
<td align="center">0.05</td>
</tr>
<tr class="even">
<td align="center"><code>returnFDR</code></td>
<td>Logical indicating whether to calculate and return the false
discovery rate (FDR).</td>
<td align="center"><code>FALSE</code></td>
</tr>
<tr class="odd">
<td align="center"><code>conditions</code></td>
<td>A character vector of length 2 specifying the conditions to compare
(e.g., <code>c("time1", "time0")</code> or
<code>c(time0_group0, time1_group1)</code> for two group designs). If
<code>NULL</code>, the comparison is inferred based on available
conditions in <code>rowData</code>.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="even">
<td align="center"><code>saveSimPath</code></td>
<td>Path to directory where simulated data sets will be saved as
<code>.rds</code> files (if desired). If <code>NULL</code>, simulated
data sets are not saved.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="odd">
<td align="center"><code>saveDePath</code></td>
<td>Path to directory where DE result files will be saved as
<code>.rds</code> files (if desired). If <code>NULL</code>, DE results
are not saved.</td>
<td align="center"><code>NULL</code></td>
</tr>
<tr class="even">
<td align="center"><code>verbose</code></td>
<td>Logical indicating whether to print progress messages during
simulation.</td>
<td align="center"><code>TRUE</code></td>
</tr>
<tr class="odd">
<td align="center"><code>...</code></td>
<td>Additional arguments passed to <code>simRescueSimDat</code>,
<code>deFunction</code>, or other internal functions.</td>
<td align="center">—</td>
</tr>
</tbody>
</table>
<p>The function returns a <code>data.frame</code> containing power (and
FDR if requested) for each simulation along with the scenario settings
and comparison conditions.</p>
</div>
<div class="section level2">
<h2 id="setting-up-the-data">Setting up the data<a class="anchor" aria-label="anchor" href="#setting-up-the-data"></a>
</h2>
<p>For this example, we will use a data set included with the
<code>rescueSim</code> package. The data set contains gene expression
data for recruited airspace macrophage (RAM) cells from bronchoalveolar
lavage samples collected from healthy adults. Data were collected from
five subjects at two time points per subject. the data set was subset to
include 1,940 genes and 970 cells. The genes included were assessed to
be invariant across time points.</p>
<p>To reduce computational cost, particularly because the differential
expression analysis in this example uses MAST with random effects, which
is computationally intensive, we further filter the data to include only
100 genes. Note that MAST does offer a parallelization option that can
speed up the analysis. Because the data set is so small, the power
estimates in this example will not be meaningful for study
planning/method benchmarking. Rather, this example is used solely to
illustrate the package functionality.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ewynn610/rescueSim" class="external-link">rescueSim</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: SummarizedExperiment</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; Loading required package: generics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'generics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     as.difftime, as.factor, as.ordered, intersect, is.element, setdiff,</span></span>
<span><span class="co">#&gt;     setequal, union</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget,</span></span>
<span><span class="co">#&gt;     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span></span>
<span><span class="co">#&gt;     rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique,</span></span>
<span><span class="co">#&gt;     unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span></span>
<span><span class="co">## Load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"RecAM_sce"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Set seed and randomly select 100 cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">24</span><span class="op">)</span></span>
<span><span class="va">RecAM_sce</span> <span class="op">&lt;-</span> <span class="va">RecAM_sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">RecAM_sce</span><span class="op">)</span>, <span class="fl">100</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co">## Estimate params from filtered data</span></span>
<span><span class="va">simParams</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/estRescueSimParams.html">estRescueSimParams</a></span><span class="op">(</span><span class="va">RecAM_sce</span>, sampleVariable <span class="op">=</span> <span class="st">"sampleID"</span>,</span>
<span>                              subjectVariable <span class="op">=</span> <span class="st">"subjectID"</span>, </span>
<span>                              timepointVariable <span class="op">=</span> <span class="st">"time"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="differential-expression-function-for-the-power-analysis">Differential expression function for the power analysis<a class="anchor" aria-label="anchor" href="#differential-expression-function-for-the-power-analysis"></a>
</h3>
<p>To run a power analysis using <code>runRescueSimPower</code>, you
must supply a custom function that performs differential expression (DE)
analysis on each simulated data set. this function should return a data
frame containing at least:</p>
<ul>
<li>
<code>gene</code>: a gene identifier</li>
<li>
<code>padj</code>: adjusted p-value (e.g. FDR adjustment)</li>
</ul>
<p>This design allows users to select a DE method appropriate for their
study design.</p>
<p>In this example, we use <code>MAST</code> with random effects to
model the hierarchical structure of our paired scRNA-seq data.
<code>MAST</code> allows the inclusion of random effects for sample and
subject, making it a reasonable choice for accounting for correlation in
our data. However, it is important to note that the performance of
<code>MAST</code> (and most other DE methods) on paired/longitudinal
single-cell data has not been fully evaluated. The results of this power
analysis should be interpreted with this limitation in mind.</p>
<p>Below is the custom DE function we will use. This function filters
data to include only genes with expression in at least
10<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>%</mi><annotation encoding="application/x-tex">\%</annotation></semantics></math>
of cells, applies log2(CPM+1) normalization, computes cellular detection
rate, fits a MAST model with random effects for sample and subject, and
returns a table with adjusted p-values for the time effect.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">deFun</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># Include only genes expressed in at least 10% of cells</span></span>
<span>    <span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">0.1</span>, <span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># Log2(CPM + 1) normalization</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">normcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="op">(</span><span class="fl">1e6</span> <span class="op">*</span> <span class="va">x</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="op">}</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Convert to SingleCellAssay</span></span>
<span>    <span class="va">sca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu">MAST</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MAST/man/FromMatrix.html" class="external-link">FromMatrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">normcounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                                             <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Calculate cellular detection rate</span></span>
<span>    <span class="va">cdr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sca</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sca</span><span class="op">)</span><span class="op">$</span><span class="va">cngeneson</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">cdr</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Fit model</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>            <span class="va">zlmCond</span> <span class="op">&lt;-</span> <span class="fu">MAST</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/MAST/man/zlm.html" class="external-link">zlm</a></span><span class="op">(</span></span>
<span>                form <span class="op">=</span> <span class="op">~</span> <span class="va">cngeneson</span> <span class="op">+</span> <span class="va">time</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">sampleID</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">subjectID</span><span class="op">)</span>,</span>
<span>                <span class="va">sca</span>,</span>
<span>                method <span class="op">=</span> <span class="st">"glmer"</span>,</span>
<span>                ebayes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                strictConvergence <span class="op">=</span> <span class="cn">TRUE</span>, silent <span class="op">=</span> <span class="cn">T</span></span>
<span>            <span class="op">)</span><span class="op">}</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Summarize LRT for time</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>            <span class="va">raw_res</span> <span class="op">&lt;-</span> <span class="fu">MAST</span><span class="fu">::</span><span class="fu">summary</span><span class="op">(</span><span class="va">zlmCond</span>, doLRT <span class="op">=</span> <span class="st">"timetime1"</span><span class="op">)</span></span>
<span>        <span class="op">}</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="va">sum_tab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">raw_res</span><span class="op">$</span><span class="va">datatable</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="co"># Keep only the H (hurdle) component rows</span></span>
<span>    <span class="va">sum_tab</span> <span class="op">&lt;-</span> <span class="va">sum_tab</span><span class="op">[</span><span class="va">sum_tab</span><span class="op">$</span><span class="va">component</span> <span class="op">==</span> <span class="st">"H"</span>, <span class="op">]</span></span>
<span>    </span>
<span>    <span class="co"># Add padj and gene columns</span></span>
<span>    <span class="va">sum_tab</span><span class="op">$</span><span class="va">padj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">sum_tab</span><span class="op">$</span><span class="va">`Pr..Chisq.`</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span></span>
<span>    <span class="va">sum_tab</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="va">sum_tab</span><span class="op">$</span><span class="va">primerid</span></span>
<span>    </span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">sum_tab</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-the-power-analysis">Running the power analysis<a class="anchor" aria-label="anchor" href="#running-the-power-analysis"></a>
</h2>
<p>We will run <code>runRescueSimPower</code> to estimate power across a
couple of illustrative scenarios. We will simulate data sets where
20<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>%</mi><annotation encoding="application/x-tex">\%</annotation></semantics></math>
of genes exhibit differential expression between two time points with a
log<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi></mi><mn>2</mn></msub><annotation encoding="application/x-tex">_2</annotation></semantics></math>
fold-change of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>±</mo><mn>0.5</mn></mrow><annotation encoding="application/x-tex">\pm0.5</annotation></semantics></math>.
We will evaluate two scenarios:</p>
<ul>
<li>Scenario 1: 3 subjects, 200 cells per sample</li>
<li>Scenario 2: 6 subjects, 100 cells per sample</li>
</ul>
<p>For simplicity, we will set the <code>minCellsPerSamp</code> and
<code>maxCellsPerSamp</code> parameters to the same value in each
scenario so that the number of cells per sample is fixed.</p>
<p>In practice, simulations like these could be used to explore
trade-offs between sequencing depth and subject recruitment. For
illustration purposes here, we will simulate only 2 data sets per
scenario and are using a very small number of genes, so the results
produced are not meaningful and cannot be used to reflect method
performance or guide study design. The goal is simply to demonstrate how
the <code>runRescueSimPower</code> function can be applied.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Update differential expression parameters in parameter object</span></span>
<span><span class="va">simParams</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/updateRescueSimParams.html">updateRescueSimParams</a></span><span class="op">(</span><span class="va">simParams</span>, </span>
<span>                                   paramValues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>propDE<span class="op">=</span><span class="fl">.2</span>, deLog2FC<span class="op">=</span><span class="fl">.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define scenarios</span></span>
<span><span class="va">scenarios</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>minCellsPerSamp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>                        maxCellsPerSamp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">100</span><span class="op">)</span>,</span>
<span>                        nSubjsPerGroup <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scenarios</span></span>
<span><span class="co">#&gt;   minCellsPerSamp maxCellsPerSamp nSubjsPerGroup</span></span>
<span><span class="co">#&gt; 1             200             200              3</span></span>
<span><span class="co">#&gt; 2             100             100              6</span></span>
<span></span>
<span><span class="co">## Set seed for reproducibility and run</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">24</span><span class="op">)</span></span>
<span><span class="va">power_res</span><span class="op">=</span><span class="fu"><a href="../reference/runRescueSimPower.html">runRescueSimPower</a></span><span class="op">(</span>baseParams <span class="op">=</span> <span class="va">simParams</span>, scenarios <span class="op">=</span> <span class="va">scenarios</span>, </span>
<span>                            deFunction <span class="op">=</span> <span class="va">deFun</span>, nSim <span class="op">=</span> <span class="fl">2</span>, returnFDR  <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running scenario 1 of 2</span></span>
<span><span class="co">#&gt;   Sim 1</span></span>
<span><span class="co">#&gt; Warning in value[[3L]](cond): DE failed for scenario 1 sim 1: Please install</span></span>
<span><span class="co">#&gt; `lme4` to use method `LMERlike`.</span></span>
<span><span class="co">#&gt;   Sim 2</span></span>
<span><span class="co">#&gt; Warning in value[[3L]](cond): DE failed for scenario 1 sim 2: Please install</span></span>
<span><span class="co">#&gt; `lme4` to use method `LMERlike`.</span></span>
<span><span class="co">#&gt; Running scenario 2 of 2</span></span>
<span><span class="co">#&gt;   Sim 1</span></span>
<span><span class="co">#&gt; Warning in value[[3L]](cond): DE failed for scenario 2 sim 1: Please install</span></span>
<span><span class="co">#&gt; `lme4` to use method `LMERlike`.</span></span>
<span><span class="co">#&gt;   Sim 2</span></span>
<span><span class="co">#&gt; Warning in value[[3L]](cond): DE failed for scenario 2 sim 2: Please install</span></span>
<span><span class="co">#&gt; `lme4` to use method `LMERlike`.</span></span></code></pre></div>
<p>The output of <code>runRescueSimPower</code> is a data frame where
each row corresponds to one simulated data set, showing the scenario
settings, simulation number, comparison conditions, power, and (if
requested) FDR. Again, with so few genes and simulations here, the
values in this example are purely illustrative and not reliable.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">power_res</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>When run with a full data set and more simulations, the output can be
summarized with power curves or other visuals to explore design
trade-offs.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SingleCellExperiment_1.30.1 SummarizedExperiment_1.38.1</span></span>
<span><span class="co">#&gt;  [3] Biobase_2.68.0              GenomicRanges_1.60.0       </span></span>
<span><span class="co">#&gt;  [5] GenomeInfoDb_1.44.0         IRanges_2.42.0             </span></span>
<span><span class="co">#&gt;  [7] S4Vectors_0.46.0            BiocGenerics_0.54.0        </span></span>
<span><span class="co">#&gt;  [9] generics_0.1.4              MatrixGenerics_1.20.0      </span></span>
<span><span class="co">#&gt; [11] matrixStats_1.5.0           rescueSim_0.99.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.1        viridisLite_0.4.2       dplyr_1.1.4            </span></span>
<span><span class="co">#&gt;  [4] vipor_0.4.7             farver_2.1.2            viridis_0.6.5          </span></span>
<span><span class="co">#&gt;  [7] fastmap_1.2.0           bluster_1.18.0          digest_0.6.37          </span></span>
<span><span class="co">#&gt; [10] rsvd_1.0.5              lifecycle_1.0.4         cluster_2.1.8.1        </span></span>
<span><span class="co">#&gt; [13] statmod_1.5.0           magrittr_2.0.3          compiler_4.5.1         </span></span>
<span><span class="co">#&gt; [16] rlang_1.1.6             sass_0.4.10             tools_4.5.1            </span></span>
<span><span class="co">#&gt; [19] igraph_2.1.4            yaml_2.3.10             data.table_1.17.6      </span></span>
<span><span class="co">#&gt; [22] knitr_1.50              dqrng_0.4.1             S4Arrays_1.8.1         </span></span>
<span><span class="co">#&gt; [25] DelayedArray_0.34.1     plyr_1.8.9              RColorBrewer_1.1-3     </span></span>
<span><span class="co">#&gt; [28] abind_1.4-8             BiocParallel_1.42.1     desc_1.4.3             </span></span>
<span><span class="co">#&gt; [31] grid_4.5.1              beachmat_2.24.0         edgeR_4.6.2            </span></span>
<span><span class="co">#&gt; [34] ggplot2_3.5.2           scales_1.4.0            gtools_3.9.5           </span></span>
<span><span class="co">#&gt; [37] MAST_1.33.0             cli_3.6.5               rmarkdown_2.29         </span></span>
<span><span class="co">#&gt; [40] crayon_1.5.3            ragg_1.4.0              metapod_1.16.0         </span></span>
<span><span class="co">#&gt; [43] httr_1.4.7              reshape2_1.4.4          scuttle_1.18.0         </span></span>
<span><span class="co">#&gt; [46] ggbeeswarm_0.7.2        cachem_1.1.0            stringr_1.5.1          </span></span>
<span><span class="co">#&gt; [49] parallel_4.5.1          XVector_0.48.0          vctrs_0.6.5            </span></span>
<span><span class="co">#&gt; [52] Matrix_1.7-3            jsonlite_2.0.0          BiocSingular_1.24.0    </span></span>
<span><span class="co">#&gt; [55] BiocNeighbors_2.2.0     ggrepel_0.9.6           irlba_2.3.5.1          </span></span>
<span><span class="co">#&gt; [58] beeswarm_0.4.0          scater_1.36.0           systemfonts_1.2.3      </span></span>
<span><span class="co">#&gt; [61] locfit_1.5-9.12         limma_3.64.1            jquerylib_0.1.4        </span></span>
<span><span class="co">#&gt; [64] glue_1.8.0              pkgdown_2.1.3           codetools_0.2-20       </span></span>
<span><span class="co">#&gt; [67] stringi_1.8.7           gtable_0.3.6            UCSC.utils_1.4.0       </span></span>
<span><span class="co">#&gt; [70] ScaledMatrix_1.16.0     tibble_3.3.0            pillar_1.10.2          </span></span>
<span><span class="co">#&gt; [73] htmltools_0.5.8.1       GenomeInfoDbData_1.2.14 R6_2.6.1               </span></span>
<span><span class="co">#&gt; [76] textshaping_1.0.1       evaluate_1.0.4          lattice_0.22-7         </span></span>
<span><span class="co">#&gt; [79] backports_1.5.0         scran_1.36.0            bslib_0.9.0            </span></span>
<span><span class="co">#&gt; [82] Rcpp_1.0.14             gridExtra_2.3           SparseArray_1.8.0      </span></span>
<span><span class="co">#&gt; [85] checkmate_2.3.2         xfun_0.52               fs_1.6.6               </span></span>
<span><span class="co">#&gt; [88] pkgconfig_2.0.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Elizabeth Wynn.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
